### DISCUSSION TITLE
What's wrong with company-clang

### DISCUSSION THREAD

**Discussion by ghost** [company-mode/company-mode#1332] [Q&A]
my emacs config uses lsp for c-mode, but when i open a c source file, company will give me an error:
```
Company backend ’company-clang’ could not be initialized:
No clause matching ‘0’
```

**Comment by dgutov** [comment]
This is probably from `company-clang--check-version`, and means that we failed to parse the version of Clang that you have installed.

What's the value of `company-clang--version`? `0`, right?

Please try debugging the function `company-clang-version`.

**Reply by ghost** [reply]
you are right, on debian 11, the clang version as below
```
$ clang --version
Debian clangd version 11.0.1-2
$ ls -l /usr/bin/clang
lrwxrwxrwx 1 root root 15 Jul 28 19:04 /usr/bin/clang -> /usr/bin/clangd
```
i have sent a pull request #1333  to fix it

**Reply by ghost** [reply]
the soft link is created by myself

**Reply by dgutov** [reply]
Hmm, but have you checked the result? Does completion work with `company-clang` now?

I'm not quite sure `clang` and `clangd` provide the same features.

**Reply by dgutov** [reply]
You can `M-x company-clang` to verify.

**Reply by ghost** [reply]
yep, i have tested, the result is ok

**Reply by dgutov** [reply]
Could you share a resulting screenshot?

I've tried this with Clangd 10 for a minute, and that failed.

**Reply by ghost** [reply]
![截屏2022-07-30 12 55 17](https://user-images.githubusercontent.com/109280366/181875362-dcc9b6fc-d051-45df-8cb3-553049bfdf24.png)

**Reply by dgutov** [reply]
A screenshot with a popup, please. And with `lsp-mode` disabled: `company-clang` is a backend that's supposed to work without LSP.

**Reply by ghost** [reply]
if i disable lsp-mode, emacs also can not complete.

**Reply by ghost** [reply]
> you are right, on debian 11, the clang version as below
> 
> ```
> $ clang --version
> Debian clangd version 11.0.1-2
> $ ls -l /usr/bin/clang
> lrwxrwxrwx 1 root root 15 Jul 28 19:04 /usr/bin/clang -> /usr/bin/clangd
> ```
> 
> i have sent a pull request #1333 to fix it

this pull request just fixs company-clang initial error

**Reply by dgutov** [reply]
Okay.

Did you see that error before you symlinked `clangd` to `clang`? Did you do that for some other purpose than to use `company-clang`?

**Reply by ghost** [reply]
after i installed clang on debian 11, the executable file is `clangd`, but in company-clang's source code, the executable is `clang`, so i created the soft link `clang` which links to `clangd`.

**Reply by ghost** [reply]
no, i don't use `company-clang` for other purposes except lsp-mode

**Reply by dgutov** [reply]
`company-clang` is not needed for `lsp-mode`. They are two separate backends with two different implementations.

So I suggest you remove that symlink, restart Emacs, and that should make the error go away.
